File: src/main/java/com/github/Doomsdayrs/Jikan4java/core/Connector.java
package com.github.Doomsdayrs.Jikan4java.core;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.Doomsdayrs.Jikan4java.enums.Days;
import com.github.Doomsdayrs.Jikan4java.enums.Season;
import com.github.Doomsdayrs.Jikan4java.enums.meta.MetaPeriod;
import com.github.Doomsdayrs.Jikan4java.enums.meta.MetaRequest;
import com.github.Doomsdayrs.Jikan4java.enums.meta.MetaType;
import com.github.Doomsdayrs.Jikan4java.types.Main.Anime.Anime;
import com.github.Doomsdayrs.Jikan4java.types.Main.Club.Club;
import com.github.Doomsdayrs.Jikan4java.types.Main.Magazine.MagazinePage;
import com.github.Doomsdayrs.Jikan4java.types.Main.Manga.Manga;
import com.github.Doomsdayrs.Jikan4java.types.Main.Person.Person;
import com.github.Doomsdayrs.Jikan4java.types.Main.Producer.ProducerPage;
import com.github.Doomsdayrs.Jikan4java.types.Main.Schedule.Day;
import com.github.Doomsdayrs.Jikan4java.types.Main.Schedule.Schedule;
import com.github.Doomsdayrs.Jikan4java.types.Main.Season.SeasonArchive.SeasonArchive;
import com.github.Doomsdayrs.Jikan4java.types.Main.Season.SeasonSearch;
import com.github.Doomsdayrs.Jikan4java.types.Main.User.User;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.IOException;
import java.util.concurrent.CompletableFuture;
//TODO Create Documentation detailing the need to convert their old deprecated Connection classes to this connector

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 * ====================================================================
 * Jikan4java
 * 16 / 05 / 2019
 *
 * @author github.com/doomsdayrs
 */
public class Connector extends Retriever {

    public Connector() {
        super(new ObjectMapper().configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true));
    }

    public Connector(OkHttpClient client) {
        super(client);
    }

    public Connector(ObjectMapper objectMapper) {
        super(objectMapper);
    }

    public Connector(JSONParser jsonParser) {
        super(jsonParser);
    }

    public Connector(Request.Builder builder) {
        super(builder);
    }

    /**
     * gets an club
     *
     * @param ID id of the club
     * @return Club
     */
    public CompletableFuture<Club> retrieveClub(int ID) {
        return retrieve(Club.class, baseURL + "/club/" + ID);
    }

    /**
     * retreives an anime
     *
     * @param ID id of the anime
     * @return Anime
     */
    public CompletableFuture<Anime> retrieveAnime(int ID) {
        return retrieve(Anime.class, baseURL + "/anime/" + ID);
    }

    /**
     * retreives an manga
     *
     * @param ID id of the manga
     * @return Manga
     */
    public CompletableFuture<Manga> retrieveManga(int ID) {
        return retrieve(Manga.class, baseURL + "/manga/" + ID);
    }

    /**
     * retreives an person
     *
     * @param ID id of the person
     * @return Person
     */
    public CompletableFuture<Person> retrievePerson(int ID) {
        return retrieve(Person.class, baseURL + "/manga/" + ID);
    }

    /**
     * retreives an character
     *
     * @param ID id of the character
     * @return Character
     */
    public CompletableFuture<Character> retrieveCharacter(int ID) {
        return retrieve(Character.class, baseURL + "/manga/" + ID);
    }


    /**
     * Returns a user object
     *
     * @param name the name of the user to retrieve
     * @return User
     * @throws IOException    IOException
     * @throws ParseException ParseException
     */
    public CompletableFuture<User> userSearch(String name) {
        return retrieve(User.class, baseURL + "/user/" + name);
    }

    /**
     * Retrieves Magazines
     *
     * @param ID   ID of magazine
     * @param page page to core for
     * @return MagazinePage object
     * @throws IOException    IOException
     * @throws ParseException ParseException
     */
    public CompletableFuture<MagazinePage> magazineSearch(int ID, int page) {
        return retrieve(MagazinePage.class, baseURL + "/magazine/" + ID + "/" + page);
    }


    /**
     * Gets meta data from API. WARNING USING MAY CAUSE ERRORS BEYOND IMAGINATION FOR ANYTHING BUT STATUS
     * @param metaRequest
     * @param metaType
     * @param metaPeriod
     * @return
     */
    public CompletableFuture getMeta(MetaRequest metaRequest, MetaType metaType, MetaPeriod metaPeriod) {
        if (metaRequest == null) return null;

        StringBuilder option = new StringBuilder();
        option.append(metaRequest);
        if (metaRequest.equals(MetaRequest.REQUEST)) {
            if (metaType != null) {
                option.append("/").append(metaType);
                if (metaPeriod != null) {
                    option.append("/").append(metaPeriod);
                } else return null;
            } else return null;
        }
        return retrieve(metaRequest.getaClass(), baseURL + "/meta/" + option.toString());
    }

    /**
     * Retrieves Producer
     *
     * @param ID   ID of magazine
     * @param page page to core for
     * @return Producer object
     * @throws IOException    IOException
     * @throws ParseException ParseException
     */
    public CompletableFuture<ProducerPage> producerSearch(int ID, int page) {
        return retrieve(ProducerPage.class, baseURL + "/producer/" + ID + "/" + page);
    }

    /**
     * Returns current schedule for all anime
     *
     * @return Schedule
     */
    public CompletableFuture<Schedule> getCurrentSchedule() {
        return retrieve(Schedule.class, baseURL + "/schedule");
    }


    /**
     * Returns all anime schedule on a certain day
     *
     * @param day Day to retrieve, Can also be other or unknown
     * @return DaySchedule object
     */
    public CompletableFuture<Day> scheduleSearch(Days day) {
        if (day != null)
            return retrieve(day.getaClass(), baseURL + "/schedule/" + day);
        return null;
    }

    /**
     * Searches for anime by season
     *
     * @param year   Year
     * @param season Season
     * @return SeasonSearchObject
     */
    public CompletableFuture<SeasonSearch> seasonSearch(int year, Season season) {
        return retrieve(SeasonSearch.class, baseURL + "/season/" + year + "/" + season);

    }

    /**
     * Returns next season of anime
     *
     * @return SeasonSearchObject
     */
    public CompletableFuture<SeasonSearch> seasonLater() {
        return retrieve(SeasonSearch.class, baseURL + "/season/later");
    }

    /**
     * Returns archive of all possible seasons and years
     *
     * @return SeasonArchive
     */
    public CompletableFuture<SeasonArchive> seasonArchive() {
        return retrieve(SeasonArchive.class, baseURL + "/season/archive");
    }

}

File: src/main/java/com/github/doomsdayrs/jikan4java/core/Connector.kt
package com.github.doomsdayrs.jikan4java.core

import com.github.doomsdayrs.jikan4java.common.JIKAN_URL
import com.github.doomsdayrs.jikan4java.data.enums.Season
import com.github.doomsdayrs.jikan4java.data.enums.meta.MetaPeriod
import com.github.doomsdayrs.jikan4java.data.enums.meta.MetaRequest
import com.github.doomsdayrs.jikan4java.data.enums.meta.MetaType
import com.github.doomsdayrs.jikan4java.data.model.main.anime.Anime
import com.github.doomsdayrs.jikan4java.data.model.main.character.Character
import com.github.doomsdayrs.jikan4java.data.model.main.club.Club
import com.github.doomsdayrs.jikan4java.data.model.main.club.ClubMemberPage
import com.github.doomsdayrs.jikan4java.data.model.main.magazine.MagazinePage
import com.github.doomsdayrs.jikan4java.data.model.main.manga.Manga
import com.github.doomsdayrs.jikan4java.data.model.main.meta.Meta
import com.github.doomsdayrs.jikan4java.data.model.main.person.Person
import com.github.doomsdayrs.jikan4java.data.model.main.producer.ProducerPage
import com.github.doomsdayrs.jikan4java.data.model.main.schedule.SchedulePage
import com.github.doomsdayrs.jikan4java.data.model.main.season.SeasonPage
import com.github.doomsdayrs.jikan4java.data.model.main.season.seasonarchive.SeasonArchivePage
import com.github.doomsdayrs.jikan4java.data.model.main.user.User
import java.util.concurrent.CompletableFuture

/*
 * This file is part of Jikan4java.
 *
 * Jikan4java is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Jikan4java is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Jikan4java.  If not, see <https://www.gnu.org/licenses/>.
 * ====================================================================
 * Jikan4java
 * 16 / 05 / 2019
 *
 * @author github.com/doomsdayrs
 */
class Connector(
	val retriever: Retriever
) {
	@Deprecated(
		"Moved to companion",
		replaceWith = ReplaceWith("SchedulePage.getSchedule(retriever)")
	)
	val currentSchedulePage: CompletableFuture<JikanResult<SchedulePage>>
		get() = SchedulePage.getSchedule(retriever)

	@Deprecated(
		"Moved to companion",
		replaceWith = ReplaceWith("Club.getByID(retriever,ID)")
	)
	fun retrieveClub(ID: Int) = Club.getByID(retriever, ID)


	@Deprecated(
		"Moved to companion",
		ReplaceWith("ClubMemberPage.get(retriever,id,1)")
	)
	@Suppress("MemberVisibilityCanBePrivate")
	fun getMembers(
		id: Int,
		page: Int
	): CompletableFuture<JikanResult<ClubMemberPage>> =
		ClubMemberPage.get(retriever, id, page)


	@Deprecated(
		"Moved to companion",
		ReplaceWith("ClubMemberPage.get(retriever,id,1)")
	)
	fun getMembers(id: Int) = getMembers(id, 1)

	@Deprecated(
		"Moved to companion",
		replaceWith = ReplaceWith("Anime.getByID(retriever,ID)")
	)
	fun retrieveAnime(ID: Int) = Anime.getByID(retriever, ID)

	@Deprecated(
		"Moved to companion",
		replaceWith = ReplaceWith("Manga.getByID(retriever,ID)")
	)
	fun retrieveManga(ID: Int) = Manga.getByID(retriever, ID)

	@Deprecated(
		"Moved to companion",
		replaceWith = ReplaceWith("Person.getByID(retriever,ID)")
	)
	fun retrievePerson(ID: Int) = Person.getByID(retriever, ID)

	@Deprecated(
		"Moved to companion",
		replaceWith = ReplaceWith("Person.getPicturesByID(retriever,id)")
	)
	fun getPersonPictures(id: Int) = Person.getPictures(retriever, id)

	@Deprecated(
		"Moved to companion",
		replaceWith = ReplaceWith("Character.getByID(retriever,ID)")
	)
	fun retrieveCharacter(ID: Int) = Character.getByID(retriever, ID)


	@Deprecated(
		"Moved to companion",
		replaceWith = ReplaceWith("Character.getPicturesByID(retriever,id)")
	)
	fun getCharacterPictures(id: Int) = Character.getPictures(retriever, id)

	@Deprecated(
		"Moved to companion",
		replaceWith = ReplaceWith("User.getByName(retriever,name)")
	)
	fun userRetrieve(name: String): CompletableFuture<JikanResult<User>> =
		User.getByName(retriever, name)


	@Deprecated(
		"Moved to companion",
		replaceWith = ReplaceWith("MagazinePage.get(retriever,ID,page)")
	)
	fun magazineSearch(
		ID: Int,
		page: Int
	): CompletableFuture<JikanResult<MagazinePage>> =
		MagazinePage.get(retriever, ID, page)

	/**
	 * Gets meta data from API.
	 * WARNING USING MAY CAUSE ERRORS BEYOND IMAGINATION FOR ANYTHING BUT STATUS
	 *
	 * @param metaRequest REQUEST / STATUS
	 * @param metaType    ANIME / CHARACTER / MANGA / PERSON / SEARCH / SCHEDULE / SEASON / TOP
	 * @param metaPeriod  MONTHLY / WEEKLY / TODAY
	 * @return Completable future of metaRequest's class
	 */
	@Deprecated("Just don't use ever, Look at the JikanStatus object and use its get")
	inline fun <reified T : Class<V>, V : Meta> getMeta(
		metaRequest: MetaRequest,
		metaType: MetaType,
		metaPeriod: MetaPeriod
	): CompletableFuture<JikanResult<T>> {
		val option = StringBuilder()
		option.append(metaRequest)

		if (metaRequest == MetaRequest.REQUEST) {
			option.append("/").append(metaType)
			option.append("/").append(metaPeriod)
		}
		return retriever("$JIKAN_URL/meta/$option")
	}

	@Deprecated(
		"Moved to companion",
		ReplaceWith(
			"ProducerPage.search(retriever,ID,page)",
			"com.github.doomsdayrs.jikan4java.data.model.main.producer.ProducerPage"
		)
	)
	fun producerSearch(
		ID: Int,
		page: Int
	) = ProducerPage.search(retriever, ID, page)


	@Deprecated(
		"Moved to companion",
		ReplaceWith("SeasonPage.get(retriever,year,season)")
	)
	fun seasonSearch(
		year: Int,
		season: Season
	): CompletableFuture<JikanResult<SeasonPage>> =
		SeasonPage.get(retriever, year, season)


	@Deprecated(
		"Moved to companion",
		ReplaceWith("SeasonPage.getNextSeason(retriever)")
	)
	fun seasonLater() = SeasonPage.getNextSeason(retriever)

	@Deprecated(
		"Moved to companion",
		ReplaceWith("SeasonArchivePage.get(retriever)")
	)
	fun seasonArchive() = SeasonArchivePage.get(retriever)
}